docs(mvux): Add List-Feed docs (#1347)
unoplatform/uno.extensions: doc/Overview/Mvux/ListFeeds.md

---
uid: Overview.Mvux.ListFeeds
---

# What are List-Feeds?

The `IListFeed<T>` is like a [Feed](xref:Overview.Mvux.Feeds) which is stateless and keeps no track of changes, but is specialized for handling collections.

In Feeds, each data request from the service, returns one single item, whereas in List-Feed each request returns a collection of items.  
Its distinctive feature is the ability to use its [Operators](xref:Overview.Mvux.ListFeeds#operators) directly on its items, rather than on the entire collection.

> Another important thing to remember is that when the service returns an empty collection, it's treated as an Empty message, and the data axis Option will be `None`, although the result was not `null`.

Reviewer: @dr1rrb is this correct - I would have assumed that if a service returns an empty collection (ie not null) the data axis would return an Option of Some with an empty collection/array

> Another important thing to remember is that when the service returns an empty collection, it's treated as an Empty message, and the data axis Option will be `None`, although the result was not `null`.

Reviewer 2: ```suggestion
Another notable characteristic of List-Feed is how it handles the absence of data. When an empty collection is returned by the service, it's treated as an Empty message. The returned data axis Option will be `None`, even though the result was not `null`.
```

Nit: This sentence explains the empty collection scenario well, but seems crowded and may be difficult to follow. I broke it down into smaller sentences.

Reviewer 2: We should also touch on the rationale for the use of `None` when both null or empty. Even if we just link to an anchor on another page, it would help to have an explanation for why this is by design.

Author: Addressed

Read this if you would like to review the [Feed Messages section](xref:Overview.Mvux.Feeds#messages).

> [!NOTE]  
> The List-Feed is using the _key equality_ to track multiple version of a same entity within different messages of the List-Feed.
(Read more about _key equality_.)[xref:Overview.KeyEquality.Concept]

## How to create a list feed

> To create an `IListFeed<T>`, use the the static class `ListFeed` to call one of the same `Async`, `AsyncEnumerable` and `Create` methods as in Feed.  

Reviewer 2: ```suggestion
To create an `IListFeed<T>`, use the static class `ListFeed` to call one of the same `Async`, `AsyncEnumerable`, and `Create` methods as in Feed.  
```

> The only difference is that it expects the Task or Async Enumerable to return an `IImmutableList<T>` instead of `T`.

Reviewer: @dr1rrb is there a reason why we've opted for requiring IImmutableList<T> rather than the much wider IEnumerable<T>?

> Here are a couple of examples of creating a List-Feed:

Reviewer 2: ```suggestion
Here are some examples of creating a List-Feed:
```

1. In this example the service returns a list of names on load/refresh - using a pull technique.

    Service code:
    ```c#
    public ValueTask<IImutableList<string>> GetNames(CancellationToken ct = default);
    ```

    Model code:
    ```c#
    public IListFeed<string> Names => ListFeed.Async(service.GetNames);
    ```
   

> 2. This one returns an immutable list of names when available - using push technique:

Reviewer: Set this to 1. to get sequential numbering?

    Service code:  
    ```c#   
    public IAsyncEnumerable<IImutableList<string>> GetNames(
        [EnumeratorCancellation] CancellationToken ct = default);
    ```

    Model code:
    ```   
    public IListFeed<string> Names => ListFeed.AsyncEnumerable(service.GetNames);
    ```

    Pull and push are explained more in the [Feeds page](xref:Overview.Mvux.Feeds#creation-of-feeds).

1. There are also two helper methods that allow you to convert from a Feed to a List-Feed and vice versa.

     - On an `IFeed<TCollection>` where `TCollection` is an `IImmutableList<TItem>`, you can call `ToListFeed()` to convert it to an `IListFeed<TItem>`.
     - In contrast, you can call `AsFeed()` on an `IListFeed<T>` to convert it to an `IFeed<IImmutableList<T>>`.

> 

Reviewer: After providing examples of how to create an IListFeed, can we have a couple of sentences talking about why use IListFeed<string> rather than IFeed<string[]>

Author: This has been explained above in lines 9-12.

## Support for pagination

There is built in support for pagination, which you can read about [here](xref:Overview.Reactive.ListFeed#paginatedasync) (Model) and [here](Overview.Reactive.InApps#pagination) (View).

## Operators

As mentioned, unlike a `Feed<List<T>>` operators on a List-Feed are directly interacting with the collection's items instead of the list itself.

### Where

This operator allows the filtering of the items.

> [!IMPORTANT]

> > If all items of the collection are filtered out, the resulting Feed will go in `None` state.

Reviewer: @dr1rrb this feels very inconsistent - I'd expect behaviour to be Some but with empty array/list/collection

---
uid: Overview.Mvux.ListFeeds
---

# What are List-Feeds?

The `IListFeed<T>` is like a [Feed](xref:Overview.Mvux.Feeds) which is stateless and keeps no track of changes, but is specialized for handling collections.

In Feeds, each data request from the service, returns one single item, whereas in List-Feed each request returns a collection of items.  
Its distinctive feature is the ability to use its [Operators](xref:Overview.Mvux.ListFeeds#operators) directly on its items, rather than on the entire collection.

Another notable characteristic of List-Feed is how it handles the absence of data. When an empty collection is returned by the service, it's treated as an Empty message. The returned data axis Option will be `None`, even though the result was not `null`.  

> The reasoning behind that is so that is that when we display a control of data items (for instance a `ListView`, and it has no items, there's no reason in displaying the `FeedView`'s `ValueTemplate` with an empty `ListView` but the "No data records" `NoneTemplate` which makes much more sense. Hence, a `null` result or an empty collection are both regarded as `None`.

Reviewer 2: ```suggestion
This is because when a control of data items is displayed with an empty collection (for instance a `ListView`), there is no reason to display the `FeedView`'s `ValueTemplate` with an empty `ListView`. The "No data records" `NoneTemplate` makes much more sense in this case. For that reason, both a `null` result and empty collection are regarded as `None`.
```

> Read this if you would like to review the [Feed Messages section](xref:Overview.Mvux.Feeds#messages).

Reviewer 2: ```suggestion
For more information, review the [Feed Messages section](xref:Overview.Mvux.Feeds#messages).
```

## How to create a list feed

To create an `IListFeed<T>`, use the static class `ListFeed` to call one of the same `Async`, `AsyncEnumerable`, and `Create` methods as in Feed.  
The only difference is that it expects the Task or Async Enumerable to return an `IImmutableList<T>` instead of `T`.

Here are some examples of creating a List-Feed:

1. In this example the service returns a list of names on load/refresh - using a pull technique.

    Service code:
    ```c#
    public ValueTask<IImutableList<string>> GetNames(CancellationToken ct = default);
    ```

    Model code:
    ```c#
    public IListFeed<string> Names => ListFeed.Async(service.GetNames);
    ```
   
2. This one returns an immutable list of names when available - using push technique:

    Service code:  
    ```c#   
    public IAsyncEnumerable<IImutableList<string>> GetNames(
        [EnumeratorCancellation] CancellationToken ct = default);
    ```

    Model code:
    ```   
    public IListFeed<string> Names => ListFeed.AsyncEnumerable(service.GetNames);
    ```

    Pull and push are explained more in the [Feeds page](xref:Overview.Mvux.Feeds#creation-of-feeds).

> 3. There are also two helper methods that allow you to convert from a Feed to a List-Feed and vice versa.

Reviewer 2: ```suggestion
3. There are also two helper methods that allow for converting from a Feed to a List-Feed and vice versa.
```

Author: ```suggestion
3. There are also two helper methods that enable conversion from a Feed to a List-Feed and vice versa.
```

>      - On an `IFeed<TCollection>` where `TCollection` is an `IImmutableList<TItem>`, you can call `ToListFeed()` to convert it to an `IListFeed<TItem>`.

Reviewer 2: ```suggestion
     - On an `IFeed<TCollection>` where `TCollection` is an `IImmutableList<TItem>`, call `ToListFeed()` to convert it to an `IListFeed<TItem>`.
```
